### PR TITLE
refactor: simplify disaggregated result

### PR DESCRIPTION
**Pull request by taharallouche**
## Description
- Simplified code complexity by using pandas built-in methods.
- Refactored the common logic in initializing the `overall` and `by_group`  inside `.create`.
- Added unit tests to cover the latter.

There are some linting changes to the doc that could be merged as part of this [other PR](https://github.com/fairlearn/fairlearn/pull/1434)

## Tests

- [ ] no new tests required
- [x] new tests added
- [ ] existing tests adjusted

## Documentation

- [ ] no documentation changes needed
- [x] user guide added or updated
- [x] API docs added or updated
- [ ] example notebook added or updated

### CODE REVIEW

File: fairlearn/metrics/_disaggregated_result.py [hunk 1]
```diff
@@ -351,49 +324,62 @@ def create(
...
-        all_grouping_names = [x for x in sensitive_feature_names]
-        if control_feature_names is not None:
-            # Note that we prepend the control feature names
-            all_grouping_names = control_feature_names + all_grouping_names
+    @staticmethod
+    def _apply_functions(
+        *,
+        data: pd.DataFrame,
+        annotated_functions: Dict[str, AnnotatedMetricFunction],
+        grouping_names: Optional[List[str]],
```
Comment (adrinjalali) [inline fairlearn/metrics/_disaggregated_result.py:346]:
```suggestion
        grouping_names: List[str] | None,
```

File: fairlearn/metrics/_disaggregated_result.py [hunk 2]
```diff
@@ -351,49 +324,62 @@ def create(
...
+    ) -> Union[pd.Series, pd.DataFrame]:
```
Comment (adrinjalali) [inline fairlearn/metrics/_disaggregated_result.py:347]:
```suggestion
    ) -> pd.Series | pd.DataFrame:
```

File: fairlearn/metrics/_disaggregated_result.py [hunk 3]
```diff
@@ -351,49 +324,62 @@ def create(
...
+        Returns
+        -------
+        Series or DataFrame
+            A Series or DataFrame with the results of the metric functions applied. If grouping_names is provided,
+            the results are grouped accordingly.
+        """
+        if grouping_names is None:
+            return apply_to_dataframe(data, metric_functions=annotated_functions)
+
+        assert len(grouping_names) > 0
```
Comment (adrinjalali) [inline fairlearn/metrics/_disaggregated_result.py:371]: we should avoid `assert` in production code. If we need to do input validation, it should raise meaningful errors with nice error messages on how to fix the issue.

**Review by adrinjalali — COMMENTED** [review]
a few nits, otherwise this looks great. Thank you so much @taharallouche .

We've also been thinking of switching to polars backend while using narwhals API internally. You might be very well qualified to take on the task.

File: fairlearn/metrics/_disaggregated_result.py (hunk 3 above)
Comment (taharallouche) [reply fairlearn/metrics/_disaggregated_result.py]: Actually an empty `grouping_names` can be a valid case (equivalent to `None`). I fixed the function and added a corresponding test case

File: fairlearn/metrics/_disaggregated_result.py (hunk 2 above)
Comment (taharallouche) [reply fairlearn/metrics/_disaggregated_result.py]: Actually, it's done :ok_hand:

**Review by adrinjalali — APPROVED** [review]
Other than the merge conflict, LGTM.

cc @riedgar-ms in case he wants to have a look.

**Review by riedgar-ms — APPROVED** [review]
Thanks!

**Comment by TamaraAtanasoska** [comment]
@taharallouche checking this first thing tomorrow again and merging thank you! I think I got confused because of the force push leaving the comments frozen so to say. because we squash and merge it you don't have to worry about how the history looks like and it doesn't matter how many commits are there, avoid to force push as it makes the review of longer PRs a bit more difficult. thanks for joining the call today too, was lovely to meet you!

**Pull request merged**